crates/nu-cli/src/commands/to_xml.rs

```diff
@@ -0,0 +1,139 @@
+            Some('p'),
+        )
+    }
+
+    fn usage(&self) -> &str {
+        "Convert table into .xml text"
+    }
+
+    async fn run(
+        &self,
+        args: CommandArgs,
+        registry: &CommandRegistry,
+    ) -> Result<OutputStream, ShellError> {
+        to_xml(args, registry).await
+    }
+}
+
+pub fn write_xml_events<W: Write>(value: &Value, writer: &mut xml::writer::EventWriter<W>) -> Result<(), ShellError> {
+    Ok(match &value.value {
+        UntaggedValue::Primitive(Primitive::Boolean(b)) => {
+            writer.write(b.to_string().as_str()).unwrap();
+        },
+        UntaggedValue::Row(o) => {
+            for (k, v) in o.entries.iter() {
+                writer.write(xml::writer::XmlEvent::start_element(k.as_str())).unwrap();
+                write_xml_events(v, writer).unwrap();
+                writer.write(xml::writer::XmlEvent::end_element()).unwrap();
+            }
+        },
+        UntaggedValue::Table(l) => {
+            writer.write(xml::writer::XmlEvent::start_element("table")).unwrap();
+            for value in l {
+                write_xml_events(value, writer).unwrap();
+            }
+            writer.write(xml::writer::XmlEvent::end_element()).unwrap();
+        },
+        _ => {
+            writer.write(xml::writer::XmlEvent::start_element("unknown")).unwrap();
+            writer.write(xml::writer::XmlEvent::end_element()).unwrap();
+        }
+    })
+}
+
+async fn to_xml(
+    args: CommandArgs,
+    registry: &CommandRegistry,
+) -> Result<OutputStream, ShellError> {
+    let registry = registry.clone();
+    let name_tag = args.call_info.name_tag.clone();
+    let (ToXMLArgs { pretty }, input) = args.process(&registry).await?;
+    let name_span = name_tag.span;
+    let input: Vec<Value> = input.collect().await;
+
+    let to_process_input = match input.len() {
+        x if x > 1 => {
+            let tag = input[0].tag.clone();
+            vec![Value {
+                value: UntaggedValue::Table(input),
+                tag,
+            }]
+        }
+        1 => input,
+        _ => vec![],
+    };
+
+    Ok(
+        futures::stream::iter(to_process_input.into_iter().map(move |value| {
+            let value_span = value.tag.span;
+
+            let indent = match &pretty {
+                Some(v) => match v.as_u64() {
+                    Ok(v) => v,
+                    _ => 0
+                },
+                _ => 0
+            };
+
+            let mut b = Vec::new();
+            let mut w = xml::writer::EmitterConfig::new()
+                .write_document_declaration(true)
+                .perform_indent(indent > 0)
+                .indent_string(" ".repeat(indent.to_usize().unwrap_or_default()))
+                .create_writer(&mut b);
+
+            match write_xml_events(&value, &mut w) {
+                Ok(xml_events) => {
+                    let s = String::from_utf8(b).unwrap();
```

Comment:
String types are confusing. Anything better than this, to avoid copies?

Reply:
If we're given bytes from the deserializer/serializer, then yeah converting them using one of the String methods is best. If this fails, we'll want to return a ShellError (basically our catch-all error type) back to the user explaining where the conversion went wrong.

Reply:
In the case of a `String` built from a procedurally-generated buffer, should I worry about invalid byte sequences?

Reply:
Most of the time I try to code defensively unless it absolutely can't happen that what's generated is bad. Generally, I see if something went wrong and return a ShellError.

With a shell, you definitely don't want the shell to panic if it possibly can, because you'll lose what you were just working on, and/or the window will close or the login will terminate.